File: 2022-04/python/841_Keys_and_Rooms.py
class Solution:

    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        # DFS Iteratively
        return self.dfs_iteratively(rooms)
        # DFS Recursively
        return self.dfs_recursively(rooms)
        # BFS
        return self.bfs(rooms)

    def dfs_iteratively(self, rooms: List[List[int]]) -> bool:
        visited = set()
        stack = [0]
        while stack:
            rn = stack.pop()
            if rn not in visited:
                visited.add(rn)
                stack += rooms[rn]
        return len(rooms) == len(visited)

    def dfs_recursively(self, rooms: List[List[int]]) -> bool:
        visited = set()
        def dfs(rn: int) -> None:
            if rn not in visited:
                visited.add(rn)
                for i in rooms[rn]:
                    dfs(i)
        dfs(0)
        return len(visited) == len(rooms)

    def bfs(self, rooms: List[List[int]]) -> bool:
        from collections import deque
        visited = set()
        queue = deque([0])
        while queue:
            rn = queue.popleft()
            if rn not in visited:
                visited.add(rn)
                queue += rooms[rn]
        return len(visited) == len(rooms)

File: 2022-04/typescript/841_Keys_and_Rooms.ts
function canVisitAllRooms(rooms: number[][]): boolean {
    // DFS Iteratively
    return dfsIteratively(rooms)
    // DFS Recursively
    return dfsRecursively(rooms)
    // BFS
    return bfs(rooms)
}

function dfsIteratively(rooms: number[][]): boolean {
    const visited = new Set<number>()
    const stack = [0]
    while (stack.length > 0) {
        const rn = stack.pop()
        if (!visited.has(rn)) {
            visited.add(rn)
            stack.push(...rooms[rn])
        }
    }
    return rooms.length === visited.size
}


function dfsRecursively(rooms: number[][]): boolean {
    function dfs(rn: number) {
        if (!visited.has(rn)) {
            visited.add(rn)
            for (let i of rooms[rn]) {
                dfs(i)
            }
        }
    }
    const visited = new Set<number>()
    dfs(0)
    return rooms.length === visited.size
}


function bfs(rooms: number[][]): boolean {
    const visited = new Set<number>()
    const queue = [0]
    while (queue.length > 0) {
        const rn = queue.shift()
        if (!visited.has(rn)) {
            visited.add(rn)
            queue.push(...rooms[rn])
        }
    }
    return rooms.length === visited.size
}